### PR TITLE
Fix: JTAG bitbanging-with-delay logic

### PR DESCRIPTION
**Pull request by dragonmux**
## Detailed description

This PR does a couple of things. The first, and most major, thing it addresses is a bug in the delayed versions of the JTAG bitbanging routines that ALTracer kindly reported on Discord.

Additionally, it addresses some cleanup that needed doing in the remote protocol implementation for the JTAG bitbanging routines, and the way that delay timing is calculated and handled as this was all quite wrong and out of whack.

The results have been tested on the LPC4370 and the GD32VF103 and show both working properly.

## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (`make PROBE_HOST=native`)
* [x] It builds as BMDA (`make PROBE_HOST=hosted`)
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

### CODE REVIEW

**Review by esden — APPROVED** [review]
LGTM

**Comment by ALTracer** [comment]
Thanks for the fix.
Re-posting the issue I described on Discord (after instructions on inserting diagnostics into low-level procedures by helpful @dragonmux) for visibility (Discord is less permanent than GitHub).

When scanning targets via BMDA hosted, it forces a frequency of 3.000MHz to the adapter. I used a `blackpill-f411ce` adapter for which naive max frequency (96M/22) was higher, 4.364MHz, and this resulted in a switch of implementations from _no_delay to _swd_delay. The _swd_delay for jtag contained a logic error that ultimately resulted in shifting out 8 least significant bits from IDCODE and a failed jtag scan (but correct 24-bit IDs in BMDA debug traces). I could reproduce this when interacting with the probe directly with GDB over ttyACM0, good scans on max freq, bad scans on lowering freq. SWD mode was not affected. Kinda counterintuitive because you'd want to reduce link frequency on seeing errors but this uncovers a lowlevel bug. Oscilloscope 2x8-bit analogue 100MHz probes on jTCK & jTDO were seeing the same (line-wise) SSPI-decoded transaction bytes on any frequencies.
Note that `swlink` build on a `bluepill` (with genuine stm32f103cbt6) defaults to 72M/22=3.273MHz and poking with BMDA/hosted did not immediately reproduce such an error. I reset the main branch backwards by like 150 commits during testing, trying to search for a regression, however swlink seemed to work and blackpill-f4 didn't.

I tested this changeset in some preliminary stage with my `blackpill-f411ce` and GD32E508 Cortex-M33 target, and the first commit in the changeset fixes the logic error. The rest of changes sets up a good infrastructure for calibrating other adapter platforms, especially non-stm32f1-based (differences in Flash ART cache, clock frequencies, GPIO APIs, platform_hwversion checks and other stuff in hot path) and brings the code to c11/c23 standards with improved readablity and scoping/const/stack efficiency.

Minor nitpick: in src/platforms/native/platform.h: in documentation/comments:
> On the _swd_delay routines

But _swd_delay was renamed to _clk_delay by the last commit in relevant source code of implementation.

File: src/platforms/native/platform.h [hunk 1]
```diff
@@ -292,6 +292,41 @@ extern bool debug_bmp;
...
+ * The 3 major JTAG bitbanging routines that get called result in these stats for
+ * clock frequency being generated with the _no_delay routines:
+ * jtag_proc.jtagtap_next(): 705.882kHz
+ * jtag_proc.jtagtap_tms_seq(): 4.4MHz
+ * jtag_proc.jtagtap_tdi_tdo_seq(): 750kHz
+ * The result is an average 1.95MHz achieved.
+ */
+#define BITBANG_NO_DELAY_FREQ 1951961U
+/*
+ * On the _swd_delay routines with the delay loops inoperative, we then get:
```
Comment (ALTracer) [inline src/platforms/native/platform.h:307]: _clk_delay

Comment (dragonmux) [reply src/platforms/native/platform.h:307]: It's noted but can't be fixed like this after the PR's already merged. It was correct at the time it was written and we just forgot to update the comment after fixing the naming.

**Pull request merged**